File: app/javascript/channels/event_chat_channel.js

import consumer from "./consumer"

document.addEventListener('turbolinks:load', () => {
  const eventElement = document.getElementById('event_chat');
  const event_id = eventElement.getAttribute('data-event-id');

  consumer.subscriptions.create({ channel: "EventChatChannel", event_id: event_id }, {
    connected() {
      // Called when the subscription is ready for use on the server
      console.log("Connected to the channel.");
    },

    disconnected() {
      // Called when the subscription has been terminated by the server
      console.log("Disconnected to the channel.");
    },

    received(data) {
      // Called when there's incoming data on the websocket for this channel
      console.log('received');
   
      const messages = document.getElementById('messages');
      messages.innerHTML += data.message;
    }
  });
});

File: app/channels/event_chat_channel.rb
# app/channels/event_chat_channel.rb

class EventChatChannel < ApplicationCable::Channel
  def subscribed
    # stream_from must be unique for each event, "event_chat_#{params[:event_id]}" is a common practice.
    stream_from "event_chat_#{params[:event_id]}"
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  def speak(data)
    # You would create a message related to an event here and broadcast it.

    message = EventMessage.create!(message: data['message'], user: current_user, event_id: params[:event_id])
    event = Event.find(params[:event_id])

    ActionCable.server.broadcast("event_chat_#{params[:event_id]}", { message: render_message(message, event) })

  end

  private

  def render_message(message, event)

    ApplicationController.renderer.render(partial: 'event_messages/message', locals: { message: message, event: event })
  end
end
